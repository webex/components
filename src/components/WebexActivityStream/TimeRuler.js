import React from 'react';
import PropTypes from 'prop-types';
import {ListSeparator} from '@momentum-ui/react';
import {format, isSameWeek, isToday, isYesterday} from 'date-fns';

import {WEBEX_COMPONENTS_CLASS_PREFIX} from '../../constants';
import './TimeRuler.scss';

/**
 * Returns a formatted timestamp based on the given date's offset from the current time.
 *
 * Divisor for messages from today display today
 * Divisor for messages from dates from a previous day display as Yesterday
 * Divisor for messages of dates from a previous day of the week (but not yesterday) display as <day of the week>
 * Divisor for messages of dates older than a week from today display as M/D/YY
 *
 * @param {Date} timestamp Date instance to format
 * @returns {string} formattedDate
 */
export function formatTimeRulerText(timestamp) {
  let formattedDate;

  if (isToday(timestamp)) {
    formattedDate = 'Today';
  } else if (isYesterday(timestamp)) {
    // Yesterday
    formattedDate = 'Yesterday';
  } else if (isSameWeek(timestamp, new Date())) {
    // Monday
    formattedDate = format(timestamp, 'iiii');
  } else {
    // 1/1/2020
    formattedDate = format(timestamp, 'P');
  }

  return formattedDate;
}

/**
 * Time Ruler component is used by the activity stream
 * to separate activities from different days.
 *
 * @param {object} props
 * @returns {object} JSX of the component
 */
export default function TimeRuler({date}) {
  const text = formatTimeRulerText(new Date(date));

  return <ListSeparator className={`${WEBEX_COMPONENTS_CLASS_PREFIX}-time-ruler`} role="listitem" text={text} />;
}

TimeRuler.propTypes = {
  date: PropTypes.string.isRequired,
};
