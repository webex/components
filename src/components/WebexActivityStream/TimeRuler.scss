@import '../../styles/variables';

.#{$WEBEX_COMPONENTS_CLASS_PREFIX}-time-ruler {
  padding: 0 3rem;
}
