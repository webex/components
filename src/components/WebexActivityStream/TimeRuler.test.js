import React from 'react';
import {
  startOfToday,
  startOfWeek,
  startOfYesterday,
  subDays,
} from 'date-fns';

import TimeRuler, {formatTimeRulerText} from './TimeRuler';

describe('Time Ruler', () => {
  let today;
  let sunday;
  let yesterday;

  beforeEach(() => {
    today = startOfToday(); // Mock date is set to Thu, Aug 1 2019 00:00
    yesterday = startOfYesterday(); // Mock date is set to Wed, July 31 2019 00:00
    sunday = startOfWeek(today); // Start of week is Sun, July 28 2019 00:00
  });

  describe('component snapshot', () => {
    test('matches with time from today', () => {
      expect(shallow(<TimeRuler date={today.toString()} />)).toMatchSnapshot();
    });

    test('matches with time from yesterday', () => {
      expect(shallow(<TimeRuler date={yesterday.toString()} />)).toMatchSnapshot();
    });

    test('matches with time from this week', () => {
      expect(shallow(<TimeRuler date={sunday.toString()} />)).toMatchSnapshot();
    });

    test('matches with older than a week time', () => {
      expect(shallow(<TimeRuler date="2019/7/26" />)).toMatchSnapshot();
    });
  });

  describe('formatTimeRulerText() returns', () => {
    test('string "Today" from today date', () => {
      expect(formatTimeRulerText(today)).toEqual('Today');
    });

    test('string "Yesterday" from yesterday date', () => {
      expect(formatTimeRulerText(yesterday)).toEqual('Yesterday');
    });

    test('name of day from a date of this week', () => {
      expect(formatTimeRulerText(sunday)).toEqual('Sunday');
    });

    test('formatted date for dates older than a week', () => {
      expect(formatTimeRulerText(subDays(sunday, 2))).toEqual('07/26/2019');
    });
  });

  afterEach(() => {
    today = null;
    yesterday = null;
    sunday = null;
  });
});
