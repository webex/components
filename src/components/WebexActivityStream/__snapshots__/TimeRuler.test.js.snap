// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Time Ruler component snapshot matches with older than a week time 1`] = `
<ListSeparator
  className="wxc-time-ruler"
  lineColor={null}
  margin={null}
  role="listitem"
  text="07/26/2019"
  textColor={null}
  textPadding={null}
/>
`;

exports[`Time Ruler component snapshot matches with time from this week 1`] = `
<ListSeparator
  className="wxc-time-ruler"
  lineColor={null}
  margin={null}
  role="listitem"
  text="Sunday"
  textColor={null}
  textPadding={null}
/>
`;

exports[`Time Ruler component snapshot matches with time from today 1`] = `
<ListSeparator
  className="wxc-time-ruler"
  lineColor={null}
  margin={null}
  role="listitem"
  text="Today"
  textColor={null}
  textPadding={null}
/>
`;

exports[`Time Ruler component snapshot matches with time from yesterday 1`] = `
<ListSeparator
  className="wxc-time-ruler"
  lineColor={null}
  margin={null}
  role="listitem"
  text="Yesterday"
  textColor={null}
  textPadding={null}
/>
`;
